utils/dcbumpgen/dcbumpgen.1
.TH DCBUMPGEN 1 "Mar 2017" "Version 1.0"
.SH NAME
dcbumpgen \- dcbumpgen a Dreamcast executable
.SH SYNOPSIS
.B dcbumpgen
.IR from
.IR to

.SH DESCRIPTION
.B dcbumpgen
is used to generate Dreamcast PVR bumpmap textures.

.SH EXAMPLES

.EX
.B
   dcbumpgen infile.png outfile.raw
.EE

.EX
.B
   dcbumpgen infile.jpg outfile.raw
.EE

.SH AUTHOR
This manual page was initially written by Stefan Galowicz <bogglez@protonmail.ch>,
for the KOS project.
.TP
The program has been initially written by Frederik Ehnbom in 2005.


doc/CHANGELOG.md
# Changelog

Platform-specific changes are prefixed with the platform name, otherwise the change is platform-independent. The name/initials of the person who is responsible for the change are listed in [square brackets] for most versions (where they are missing from early versions, assume that they should be attributed to Megan Potter).

## KallistiOS version 2.2.2
- **Dreamcast**: Add missing header to `snd_stream.h` [Donald Haase ==DH]
- Fix bug causing erratic behavior in `readdir` [Paul Cercueil == PC]
- No longer build GDB by default with toolchain [Eric Fradella == EF]
- Fix building older toolchains with newer host GCC [PC]
- Correct cmake include path for addons [DH]
- **Dreamcast**: Prevent controller from unregistering if sub-devices still active [PC]
- Improved support for C static asserts [Falco Girgis == FG][PC]
- Correct errors in doxygen processing [DH][PC]
- Improve thread safety of fs operations [PC]
- Fix CLion compatibility with kos-cc wrappers [kouta]

## KallistiOS version 2.2.1
- **Dreamcast**: Fine-tune timer rate to avoid clock drift [Paul Cercueil == PC]
- **Dreamcast**: Update keyboard LED/mod status handling, add new example to demonstrate [Troy Davis == GPF]
- **Dreamcast**: Detect 3rd party memory cards and disable sending screen/clock commands to them [Donald Haase == DH]
- **Dreamcast**: Support for 1st party rumble packs [DH]
- **Dreamcast**: Various fixes and expansions to the cmake environment [Falco Girgis == FG][PC]

## KallistiOS version 2.2.0
- Added . & .. directories to filesystems that lack it [Andress Barajas == AB]
- Replaced previous implementation of realpath() to remove license from AUTHORS [AB]
- Made FD_SETSIZE configurable in kos/opts.h [AB] 
- **Dreamcast**: Enabled hybrid PVR DR/DMA vertex submission in driver + sped up pvr_prim() [Falco Girgis == FG]
- Add thread priority boosting system [Paul Cercueil == PC]
- **Dreamcast**: Use DMA for iso9660 reading [PC]
- **Dreamcast**: Dynamic maple device allocation [PC]
- Add performance monitor API [PC]
- Add/Fixed stat() implementations for all filesystems [AB]
- **Dreamcast**: Add pvr palette example [AB]
- **Dreamcast**: Fixed, expanded, and created example for library loading [Ruslan Rostovtsev == RR]
- **Dreamcast**: Cleaned up, documented, and enhanced BIOS font API [FG]
- Expanded KOS_INIT_FLAG options to include default VFSes and library loading [FG]
- **Dreamcast**: Rework PVR hybrid mode + IRQ handling [PC]
- Removed defines for BSD-style types (u_char, u_int, etc) [Donald Haase == DH]
- **Dreamcast**: Add support and update toolchain profiles for Newlib 4.5.0, Binutils 2.43.1, and GDB 15.2 [Eric Fradella == EF]
- **Dreamcast**: Make m4-single the default floating-point ABI [EF]
- **Dreamcast**: Allow building a toolchain for the m4-nofpu ABI [DH]
- **Dreamcast**: Add DMAC API [PC]
- **Dreamcast**: Add basic MMU functionality [PC]
- **Dreamcast**: vmu_pkg loading from .ico icons [PC]
- **Dreamcast**: vmu_pkg Manage file headers outside of standard fs [PC]
- **Dreamcast**: Added Raytris raylib example by Niisoks [FG]
- Added support for pthread_condattr_setclock() [FG]
- Expanded supported sysconf queries [FG]
- **Dreamcast**: iso9660 multi-sector reads to use DMA and no longer cached [RR]
- **Dreamcast**: Substantial cleanups and refactoring for PVR and IRQ APIs [FG]
- Various internal adjustments to support non-Dreamcast platforms [DH]
- **Dreamcast**: maple now allows per-driver device attach/detach callbacks [PC]
- **Dreamcast**: iso9660 file table now dynamic, preventing the need for manual size adjustment [PC]
- **Dreamcast**: cdrom add support for irq-based DMA reads as well as data streaming [RR]
- Removed old pthreads implementation from kernel/libc [Lawrence Sebald == LS]
- Added new (more complete/compliant) pthreads implementation in addons/libpthread [LS]
- Added implementation of thread barriers [LS]
- Added support for pthread_condattr_setclock() [FG]
- **Dreamcast**: maple Keyboard driver significantly rewritten, new event driven API added [FG]
- **Dreamcast**: Add SCI driver [RR]
- **Dreamcast**: Update SD card driver for SCI support and optional CRC check disabling [RR]
- New macro __is_defined() which should be used in place of \#ifdef guards internally [PC]

## KallistiOS version 2.1.1
- Added pvrtex utility by TapamN to utils [Daniel Fairchild == DF]
- **Dreamcast**: Added a set of known working purupuru bit patterns and a browsing mechanism to the rumble example [DF]
- **Dreamcast**: Added new example to demonstrate the use of the PVR to draw lines with quads (pvr/pvrline) [Jason Martin == JM]
- **Dreamcast**: Add network speedtest example [Andress Barajas == AB]

## KallistiOS version 2.1.0
- Cleaned up generated stubs files on a make clean [Lawrence Sebald == LS]
- **Dreamcast**: Added a function to detect if the program is being run on a retail Dreamcast or a Set5.xx device [LS]
- **Dreamcast**: Fixed an issue with the SIP driver that would cause programs to freeze if a microphone was connected at program startup [LS]
- **Dreamcast**: Rearranged the SIP driver to get rid of the internal buffer [LS]
- Removed the GBA, ia32, and ps2 ports. If someone really wants them back and wants to maintain them, let me know. [LS]
- **Dreamcast**: Fixed fmath functions so they compile in C99 mode [LS]
- Cleaned up warnings with -std=gnu99 for C code [LS]
- Cleaned up warnings with -std=c99 for C code [LS]
- **Dreamcast**: Changed the PVR render-to-texture support to allow for two render-to-texture passes without an intervening render-to-screen pass [LS]
- Fixed an issue in fs_ext2 that would allow you to rename a directory to be a child of itself, thus chopping it off from the rest of the directory tree [LS]
- Added a fs_readlink() function to resolve symlinks [LS]
- Fixed the vqenc and kmgenc utilities for compilation on a 64-bit OS [LS]
- **Dreamcast**: Corrected an issue with video cable detection with GCC 4.8.x [LS]

utils/dcbumpgen/dcbumpgen.c
/*   
   dcbumpgen - creates polar angles to be used for bumpmapping
   Copyright (c) 2005 Fredrik Ehnbom

   This software is provided 'as-is', without any express or implied 
   warranty. In no event will the authors be held liable for any 
   damages arising from the use of this software.

   Permission is granted to anyone to use this software for any 
   purpose, including commercial applications, and to alter it and 
   redistribute it freely, subject to the following restrictions:

   1. The origin of this software must not be misrepresented; you 
      must not claim that you wrote the original software. If you use
      this software in a product, an acknowledgment in the product 
      documentation would be appreciated but is not required.

   2. Altered source versions must be plainly marked as such, and 
      must not be misrepresented as being the original software.

   3. This notice may not be removed or altered from any source 
      distribution.
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

/* TODO:
 * With this utility included, there are now three utilities
 * (the other two being "vqenc" and "kmgenc") using common texture
 * operations such as twiddling and loading of png and jpeg images.
 * Maybe it is time to break that stuff out into its own library
 * or create a megatool consisting of all the tree tools in one
 * executable.
 */
#include "get_image.h"

static void printUsage() {
	printf("dcbumpgen - Dreamcast bumpmap generator v0.1\n");
	printf("Copyright (c) 2005 Fredrik Ehnbom\n");
	printf("usage: dcbumpgen <infile.png/.jpg> <outfile.raw>\n");
}

static int isPowerOfTwo(unsigned x)
{
	return x && !(x & (x-1));
}

/* twiddling stuff copied from kmgenc.c */
#define TWIDTAB(x) ( (x&1)|((x&2)<<1)|((x&4)<<2)|((x&8)<<3)|((x&16)<<4)| \
	((x&32)<<5)|((x&64)<<6)|((x&128)<<7)|((x&256)<<8)|((x&512)<<9) )
#define TWIDOUT(x, y) ( TWIDTAB((y)) | (TWIDTAB((x)) << 1) )
#define MIN(a, b) ( (a)<(b)? (a):(b) )

#define ERROR(...) { ret = 1; fprintf(stderr, __VA_ARGS__); goto cleanup; }

int main(int argc, char **argv) {
	int ret = 0;
	int y, x;
	int imgpos, dest;
	image_t img = {0};
	FILE *fp = 0;
	unsigned char *buffer = 0;
	short * twidbuffer    = 0;

	if (argc != 3) {
		printUsage();
		return 0;
	}

	if (get_image(argv[1], &img) < 0) {
		ERROR("Cannot open %s\n", argv[1]);
	}

	fp = fopen(argv[2], "wb");
	if(!fp) {
		ERROR("Cannot open file %s!\n", argv[2]);
	}

	buffer = malloc(2 * img.w * img.h);
	if(!buffer) {
		ERROR("Cannot allocate memory for image data!\n");
	}

	twidbuffer = malloc(2 * img.w * img.h);
	if(!buffer) {
		ERROR("Cannot allocate memory for twiddle buffer!\n");
	}

	if(!isPowerOfTwo(img.w) || !isPowerOfTwo(img.h)) {
		ERROR("Image dimensions %ux%u are not a power of two!\n", img.w, img.h);
	}

	imgpos = 1; /* 1 to skip the alpha-channel */
	dest = 0;
	for (y = 0; y < img.h; y++) {
		for (x = 0; x < img.w; x++, imgpos += 4) {
			double diffy = 0;
			double diffx = 0;
			if (y > 0 && x > 0) {
				diffy = (img.data[imgpos - img.stride] - img.data[imgpos]) / 255.0;
				diffx = (img.data[imgpos - 4] - img.data[imgpos]) / 255.0;
			}

			/* Rotation = R
			   0 -> almost 360 degrees */
			double rot = atan2(diffy, diffx);
			int rotation = (int) ((rot / (2 * 3.1415927)) * 255);

			/* Elevation = S
			   0 -> almost 90 degrees */
			int elevation = (int) (255 * (1 - fabs(diffx) - fabs(diffy)));
			if (elevation < 0) elevation = 0;

			buffer[dest] = rotation; dest++;
			buffer[dest] = elevation; dest++;
		}
	}

	/* twiddle code based on code from kmgenc.c */
	int min = MIN(img.w, img.h);
	int mask = min-1;
	short *sbuffer = (short*) buffer;
	for (y=0; y<img.h; y++) {
		int yout = y;
		for (x=0; x<img.w; x++) {
			twidbuffer[TWIDOUT(x&mask, yout&mask) +
				(x/min + yout/min)*min*min] = sbuffer[y*img.w+x];
		}
	}

	if(fwrite(twidbuffer, 2* img.w * img.h, 1, fp) != 1) {
		ERROR("Cannot write twiddle buffer!\n");
	}

cleanup:
	if(fp) fclose(fp);
	if(buffer) free(buffer);
	if(twidbuffer) free(twidbuffer);
	if(img.data) free(img.data);

	return ret;
}

